## Final Project Submission

Please fill out:
* Student name: Clyde Ochieng
* Student pace:  full time
* Scheduled project review date/time: 
* Instructor name: Nikita Njoroge 
* Blog post URL:


# MICROSOFT MOVIE STUDIO EXPLORATION AND DATA ANALYSIS

## 1. Business Understanding

### a) Introduction 
Microsoft is venturing into the entertainment industry by establishing a new movie studio to create original video content. The company recognizes the success of big players in the industry and aims to capitalize on this trend. However, lacking experience in movie production, Microsoft needs insights into the current landscape of successful films at the box office. The goal is to inform strategic decisions regarding the types of films that will resonate well with audiences and maximize profitability.

### b) Problem Statement
Microsoft wants to succeed in the entertainment sector by navigating the complexities of the movie studio business by launching a new venture that will produce original video content that will appeal to viewers and yield maximum profit. Microsoft understands the necessity to learn about the present world of box office-performing films, as it has no past expertise in film creation. The strategic decisions the studio needs to make about the kinds of movies it should make to attract viewers and survive in the cutthroat industry depend heavily on this information.

### c) Main Objective
To empower Microsoft's new movie studio with actionable insights into the dynamics of the film industry, particularly focusing on box office performance by leveraging data analysis and exploration techniques using pandas, where the goal is to identify the key factors contributing to the success of movies, including genres, talented writers, directors, and popularity metrics.

### d) Specific Objectives

1. To explore Box Office Performance by analysing historical data on movies, focusing on key metrics such as total gross, average rating, and number of votes to understand box office performance.

2. Translate insights into recommendations by Providing actionable recommendations for Microsoft's movie studio, suggesting specific genres, writers, directors, and popularity metrics to consider for future movie productions.

3. Position for success by strategically positioning the new movie studio to create content that not only achieves financial success but also resonates with a diverse audience, fostering long-term growth and success in the entertainment market. 


### e) Experimental Design
1. Data Collection
2. Read and check the data
3. Cleaning the data
4. Exploratory Data Analysis
5. Conclusions and Recommendations


### f) Data Understanding 
The data used in this project are data collected from:
1. https://www.boxofficemojo.com/
2. https://www.imdb.com/
3. https://www.rottentomatoes.com/
4. https://www.themoviedb.org/
5. https://www.the-numbers.com/

The imdb dataset contained 8 tables:
|No.| Table    |
|---| ---       |
|1|movie_basics|
|2|directors|
|3|known_for|
|4|movie_akas|
|5|movie_ratings|
|6|persons|
|7|principals|
|8|writers|

## 2. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import csv
import os
import sqlite3
import matplotlib.pyplot as plt
%matplotlib inline
import zipfile

## 3. Reading and merging the Data

### 3a. IMDB Movies

### i)Reading the data

In [2]:
# Extract IMDb SQL .db file
with zipfile.ZipFile('./zippedData/im.db.zip') as zipObj:
    # Extract all contents of .zip file into current directory
    zipObj.extractall(path='./zippedData/')
    
# Create connection to IMDb DB
con = sqlite3.connect('./zippedData/im.db')
#import data from rt.movie_info.tsv.gz
movie_info = pd.read_csv('./zippedData/rt.movie_info.tsv.gz',sep="\t")
#rt.reviews.tsv.gz
reviews = pd.read_csv('./zippedData/rt.reviews.tsv.gz',sep="\t", encoding = 'unicode_escape')
#bom.movie_gross.csv.gz
movie_gross = pd.read_csv('./zippedData/bom.movie_gross.csv.gz')
#tmdb.movies.csv.gz
tmdb = pd.read_csv('./zippedData/tmdb.movies.csv.gz')
#tn.movie_budgets.csv.gz
budgets = pd.read_csv('./zippedData/tn.movie_budgets.csv.gz')

In [3]:
im = os.path.join('im.db')
# Open up a connection
conn = sqlite3.connect(im)
# Initialize a cursor
cursor = conn.cursor()

In [4]:
import pandas as pd
import zipfile
import sqlite3

# Path to the zip file containing the SQLite database
zip_file_path = r'C:\Users\user\Downloads\Moringa\dsc-phase-1-project-v2-4\zippedData\im.db.zip'

# Extract the zip file
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    # Assuming there's only one file in the zip, get its name
    file_name_inside_zip = zip_ref.namelist()[0]
    # Extract the file
    zip_ref.extractall(r'C:\Users\user\Downloads\Moringa\dsc-phase-1-project-v2-4\unzippedData')

# Connect to the SQLite database
conn = sqlite3.connect(r'C:\Users\user\Downloads\Moringa\dsc-phase-1-project-v2-4\unzippedData\im.db')
# Viewing the list of tables in the IMDb database
imdb_tables = pd.read_sql("""
SELECT name AS TableNames
FROM sqlite_master
WHERE type='table'
""", conn)

# Print the DataFrame
print(imdb_tables)



      TableNames
0   movie_basics
1      directors
2      known_for
3     movie_akas
4  movie_ratings
5        persons
6     principals
7        writers


In [5]:
#previewing the directors table
q = pd.read_sql("""
SELECT *
FROM directors
;
""",conn)
q.head()

movie_id  person_id
0  tt0285252  nm0899854
1  tt0462036  nm1940585
2  tt0835418  nm0151540
3  tt0835418  nm0151540
4  tt0878654  nm0089502

In [6]:
#previewing movie_ratings table
q1 = pd.read_sql("""
SELECT *
FROM movie_ratings
;
""",conn)
q1.head()

movie_id  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [7]:
#previewing movie_basics table
q2 = pd.read_sql("""
SELECT *
FROM movie_basics
;
""",conn)
q2.head()

movie_id                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [8]:
#previewing the persons table
q3 = pd.read_sql("""
SELECT *
FROM persons
;
""",conn)
q3.head()

person_id       primary_name  birth_year  death_year  \
0  nm0061671  Mary Ellen Bauder         NaN         NaN   
1  nm0061865       Joseph Bauer         NaN         NaN   
2  nm0062070         Bruce Baum         NaN         NaN   
3  nm0062195       Axel Baumann         NaN         NaN   
4  nm0062798        Pete Baxter         NaN         NaN   

                                 primary_profession  
0         miscellaneous,production_manager,producer  
1        composer,music_department,sound_department  
2                        miscellaneous,actor,writer  
3  camera_department,cinematographer,art_department  
4  production_designer,art_department,set_decorator

### ii)Merging movie_basics to movie_ratings, directors and persons table.

In [9]:
#joining movie_basics to movie_ratings on movie_id and previewing the top
imdb = pd.read_sql("""
SELECT *
FROM movie_basics b
JOIN movie_ratings r
USING (movie_id)
JOIN directors d
USING (movie_id)
JOIN persons p
USING (person_id)
;
""",conn)
imdb.head()

movie_id                    primary_title   original_title  start_year  \
0  tt0063540                        Sunghursh        Sunghursh        2013   
1  tt0063540                        Sunghursh        Sunghursh        2013   
2  tt0063540                        Sunghursh        Sunghursh        2013   
3  tt0063540                        Sunghursh        Sunghursh        2013   
4  tt0066787  One Day Before the Rainy Season  Ashad Ka Ek Din        2019   

   runtime_minutes              genres  averagerating  numvotes  person_id  \
0            175.0  Action,Crime,Drama            7.0        77  nm0712540   
1            175.0  Action,Crime,Drama            7.0        77  nm0712540   
2            175.0  Action,Crime,Drama            7.0        77  nm0712540   
3            175.0  Action,Crime,Drama            7.0        77  nm0712540   
4            114.0     Biography,Drama            7.2        43  nm0002411   

          primary_name  birth_year  death_year        primary_profession  
0  Harnam Singh Rawail      1921.0      2004.0  director,writer,producer  
1  Harnam Singh Rawail      1921.0      2004.0  director,writer,producer  
2  Harnam Singh Rawail      1921.0      2004.0  director,writer,producer  
3  Harnam Singh Rawail      1921.0      2004.0  director,writer,producer  
4            Mani Kaul      1944.0      2011.0     director,writer,actor

In [10]:
#closing the connection to the database
conn.close()

In [11]:
#viewing the bottom of the imdb dataframe
imdb.tail()

movie_id               primary_title              original_title  \
181382  tt9914642                   Albatross                   Albatross   
181383  tt9914642                   Albatross                   Albatross   
181384  tt9914942  La vida sense la Sara Amat  La vida sense la Sara Amat   
181385  tt9914942  La vida sense la Sara Amat  La vida sense la Sara Amat   
181386  tt9916160                  Drømmeland                  Drømmeland   

        start_year  runtime_minutes       genres  averagerating  numvotes  \
181382        2017              NaN  Documentary            8.5         8   
181383        2017              NaN  Documentary            8.5         8   
181384        2019              NaN         None            6.6         5   
181385        2019              NaN         None            6.6         5   
181386        2019             72.0  Documentary            6.5        11   

        person_id        primary_name  birth_year  death_year  \
181382  nm5300859        Chris Jordan         NaN         NaN   
181383  nm5300859        Chris Jordan         NaN         NaN   
181384  nm1716653           Laura Jou         NaN         NaN   
181385  nm1716653           Laura Jou         NaN         NaN   
181386  nm5684093  Joost van der Wiel         NaN         NaN   

                     primary_profession  
181382           director,writer,editor  
181383           director,writer,editor  
181384   miscellaneous,actress,director  
181385   miscellaneous,actress,director  
181386  director,cinematographer,writer

## c)Calculating the count of entries in the IMDb dataframe

In [12]:
#determining the number of records in the imdb dataframe
imdb.shape

(181387, 13)

In [13]:
#checking the information
imdb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181387 entries, 0 to 181386
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   movie_id            181387 non-null  object 
 1   primary_title       181387 non-null  object 
 2   original_title      181387 non-null  object 
 3   start_year          181387 non-null  int64  
 4   runtime_minutes     163584 non-null  float64
 5   genres              180047 non-null  object 
 6   averagerating       181387 non-null  float64
 7   numvotes            181387 non-null  int64  
 8   person_id           181387 non-null  object 
 9   primary_name        181387 non-null  object 
 10  birth_year          54805 non-null   float64
 11  death_year          1342 non-null    float64
 12  primary_profession  181262 non-null  object 
dtypes: float64(4), int64(2), object(7)
memory usage: 18.0+ MB


### d)checking out the stats summary of the IMDb data.

In [14]:
#previewing the summary statistics of imdb 
imdb.describe()

start_year  runtime_minutes  averagerating      numvotes  \
count  181387.000000    163584.000000  181387.000000  1.813870e+05   
mean     2014.309802        97.789484       6.217683  4.955524e+03   
std         2.536111       194.434689       1.388026  3.760931e+04   
min      2010.000000         3.000000       1.000000  5.000000e+00   
25%      2012.000000        84.000000       5.400000  1.900000e+01   
50%      2014.000000        94.000000       6.300000  6.600000e+01   
75%      2016.000000       107.000000       7.200000  3.110000e+02   
max      2019.000000     51420.000000      10.000000  1.841066e+06   

         birth_year   death_year  
count  54805.000000  1342.000000  
mean    1969.097856  2014.908346  
std       12.499740     4.866581  
min     1870.000000  1944.000000  
25%     1962.000000  2014.000000  
50%     1971.000000  2016.000000  
75%     1978.000000  2018.000000  
max     2004.000000  2019.000000

## Observations:
1. There are columns with missing data, these are runtime_minutes, genres, birth_year, death_year and primary_profession.
2. runtime_minutes has a movie that is 51420 minutes and 3 minutes long and the standard deviation is at 194.4.
3. The genres are separated by commas.

### 3b. Bom_movies 

In [15]:
import pandas as pd

# Read the CSV file
movie_gross = pd.read_csv('./zippedData/bom.movie_gross.csv.gz')

# Print DataFrame in the specified format with index
bom_movies.head() 



NameError: name 'bom_movies' is not defined

In [ ]:
import pandas as pd

# Read the CSV file
bom_movies = pd.read_csv('./zippedData/bom.movie_gross.csv.gz')

# Print the bottom of the DataFrame in the specified format with index
bom_movies.tail()


title      studio  domestic_gross foreign_gross  \
3382                    The Quake       Magn.          6200.0           NaN   
3383  Edward II (2018 re-release)          FM          4800.0           NaN   
3384                     El Pacto        Sony          2500.0           NaN   
3385                     The Swan  Synergetic          2400.0           NaN   
3386            An Actor Prepares       Grav.          1700.0           NaN   

      year  
3382  2018  
3383  2018  
3384  2018  
3385  2018  
3386  2018

In [ ]:
import pandas as pd

# Read the CSV file
bom_movies = pd.read_csv('./zippedData/bom.movie_gross.csv.gz')
#determining the number of records in the bom_movies dataframe
bom_movies.shape

(3387, 5)

In [ ]:
import pandas as pd

# Read the CSV file
bom_movies = pd.read_csv('./zippedData/bom.movie_gross.csv.gz')

# Print DataFrame information
bom_movies.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB
None


In [ ]:
#previewing the summary statistics of bom_movies
bom_movies.describe()

domestic_gross         year
count    3.359000e+03  3387.000000
mean     2.874585e+07  2013.958075
std      6.698250e+07     2.478141
min      1.000000e+02  2010.000000
25%      1.200000e+05  2012.000000
50%      1.400000e+06  2014.000000
75%      2.790000e+07  2016.000000
max      9.367000e+08  2018.000000

### Observations:
1. bom_movies has the following columns with missing values: domestic_gross and foreign_gross.
2. The foreign_gross column is in string format.



### 3c. Rt_movies

In [ ]:
import pandas as pd

# Read the CSV files
reviews = pd.read_csv('./zippedData/rt.reviews.tsv.gz', sep="\t", encoding='unicode_escape')
movie_info = pd.read_csv('./zippedData/rt.movie_info.tsv.gz', sep="\t")

# Display the top of the movie_info DataFrame
movie_info.head()



id                                           synopsis rating  \
0   1  This gritty, fast-paced, and innovative police...      R   
1   3  New York City, not-too-distant-future: Eric Pa...      R   
2   5  Illeana Douglas delivers a superb performance ...      R   
3   6  Michael Douglas runs afoul of a treacherous su...      R   
4   7                                                NaN     NR   

                                 genre          director  \
0  Action and Adventure|Classics|Drama  William Friedkin   
1    Drama|Science Fiction and Fantasy  David Cronenberg   
2    Drama|Musical and Performing Arts    Allison Anders   
3           Drama|Mystery and Suspense    Barry Levinson   
4                        Drama|Romance    Rodney Bennett   

                            writer  theater_date      dvd_date currency  \
0                   Ernest Tidyman   Oct 9, 1971  Sep 25, 2001      NaN   
1     David Cronenberg|Don DeLillo  Aug 17, 2012   Jan 1, 2013        $   
2                   Allison Anders  Sep 13, 1996  Apr 18, 2000      NaN   
3  Paul Attanasio|Michael Crichton   Dec 9, 1994  Aug 27, 1997      NaN   
4                     Giles Cooper           NaN           NaN      NaN   

  box_office      runtime             studio  
0        NaN  104 minutes                NaN  
1    600,000  108 minutes  Entertainment One  
2        NaN  116 minutes                NaN  
3        NaN  128 minutes                NaN  
4        NaN  200 minutes                NaN

In [ ]:
import pandas as pd

# Read the CSV files
reviews = pd.read_csv('./zippedData/rt.reviews.tsv.gz', sep="\t", encoding='unicode_escape')
movie_info = pd.read_csv('./zippedData/rt.movie_info.tsv.gz', sep="\t")

# Display the bottom of the movie_info DataFrame
movie_info.tail()


id                                           synopsis rating  \
1555  1996  Forget terrorists or hijackers -- there's a ha...      R   
1556  1997  The popular Saturday Night Live sketch was exp...     PG   
1557  1998  Based on a novel by Richard Powell, when the l...      G   
1558  1999  The Sandlot is a coming-of-age story about a g...     PG   
1559  2000  Suspended from the force, Paris cop Hubert is ...      R   

                                                  genre            director  \
1555   Action and Adventure|Horror|Mystery and Suspense                 NaN   
1556                 Comedy|Science Fiction and Fantasy        Steve Barron   
1557  Classics|Comedy|Drama|Musical and Performing Arts      Gordon Douglas   
1558    Comedy|Drama|Kids and Family|Sports and Fitness  David Mickey Evans   
1559  Action and Adventure|Art House and Internation...                 NaN   

                                                writer  theater_date  \
1555                                               NaN  Aug 18, 2006   
1556  Terry Turner|Tom Davis|Dan Aykroyd|Bonnie Turner  Jul 23, 1993   
1557                                               NaN   Jan 1, 1962   
1558                  David Mickey Evans|Robert Gunter   Apr 1, 1993   
1559                                        Luc Besson  Sep 27, 2001   

          dvd_date currency  box_office      runtime             studio  
1555   Jan 2, 2007        $  33,886,034  106 minutes    New Line Cinema  
1556  Apr 17, 2001      NaN         NaN   88 minutes  Paramount Vantage  
1557  May 11, 2004      NaN         NaN  111 minutes                NaN  
1558  Jan 29, 2002      NaN         NaN  101 minutes                NaN  
1559  Feb 11, 2003      NaN         NaN   94 minutes  Columbia Pictures

In [ ]:
import pandas as pd

# Read the CSV files
reviews = pd.read_csv('./zippedData/rt.reviews.tsv.gz', sep="\t", encoding='unicode_escape')
movie_info = pd.read_csv('./zippedData/rt.movie_info.tsv.gz', sep="\t")

# Determine the number of records in the movie_info DataFrame
movie_info.shape



(1560, 12)

In [ ]:
import pandas as pd

# Read the CSV files
reviews = pd.read_csv('./zippedData/rt.reviews.tsv.gz', sep="\t", encoding='unicode_escape')
movie_info = pd.read_csv('./zippedData/rt.movie_info.tsv.gz', sep="\t")

# Print information about the reviews DataFrame
reviews.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54432 entries, 0 to 54431
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          54432 non-null  int64 
 1   review      48869 non-null  object
 2   rating      40915 non-null  object
 3   fresh       54432 non-null  object
 4   critic      51710 non-null  object
 5   top_critic  54432 non-null  int64 
 6   publisher   54123 non-null  object
 7   date        54432 non-null  object
dtypes: int64(2), object(6)
memory usage: 3.3+ MB


### Observations:
1. The columns review,rating and critic have missing values
2. rt_reviews can be merged with rt_movies on id as they are from the same source.
3. There are multiple reviews per movie.
4. The rating column appears to be out of 5.

### Objective 2 
### Clean the data 

### 6. Data Cleaning
This section deals with data cleaning methods that prepares the data for exploratory analysis. It tackles the problems with the datasets that have been observed while exploring them and any other that occur afterwards. It addresses the following:
   - Duplicated rows
   - Missing values
   - Unwanted observations
   - Structural issues
   - Outliers
   - Invalid data
   - Merges
   
### 6.1. Duplicate Rows
It is helpful to define a function that checks for duplicate rows in all the dataframes.   

In [16]:
#defining a function checking for number of duplicated rows
def check_duplicates(df):
    duplicates = df.duplicated().sum()
    print (f"There are {duplicates} duplicated rows")

In [17]:
#checking for duplicates in imdb dataframe
check_duplicates(imdb)

There are 95357 duplicated rows
